File: p11-platform/apps/web/utils/services/url-safety.ts
/**
 * Shared SSRF-safe URL validation for user-supplied URLs that will be
 * stored and later fetched by the data-engine or other services.
 */

const APARTMENTS_COM_HOSTS = new Set(['apartments.com', 'www.apartments.com'])

function parseUrl(value: unknown): URL | null {
  if (typeof value !== 'string' || value.trim().length === 0) {
    return null
  }
  try {
    return new URL(value.trim())
  } catch {
    return null
  }
}

function isPrivateOrReservedIp(host: string): boolean {
  // IPv4 literal checks
  const ipv4 = host.match(/^(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})$/)
  if (ipv4) {
    const octets = ipv4.slice(1).map(Number)
    if (octets.some((o) => o > 255)) return true
    const [a, b] = octets
    return (
      a === 0 ||
      a === 10 ||
      a === 127 ||
      (a === 169 && b === 254) ||
      (a === 172 && b >= 16 && b <= 31) ||
      (a === 192 && b === 168) ||
      a >= 224
    )
  }
  // IPv6 literals (URL hostname wraps them in brackets already stripped by URL parser)
  if (host.includes(':')) {
    const normalized = host.replace(/^\[|\]$/g, '').toLowerCase()
    return (
      normalized === '::' ||
      normalized === '::1' ||
      normalized.startsWith('fe80') ||
      normalized.startsWith('fc') ||
      normalized.startsWith('fd')
    )
  }
  return false
}

/**
 * True only for http(s) URLs pointing at public hostnames.
 * Rejects localhost, private/link-local/reserved IPs, bare internal
 * hostnames, and URLs with embedded credentials.
 */
export function isSafePublicHttpUrl(value: unknown): boolean {
  const url = parseUrl(value)
  if (!url) return false
  if (url.protocol !== 'http:' && url.protocol !== 'https:') return false
  if (url.username || url.password) return false
  const host = url.hostname.toLowerCase()
  if (!host || host === 'localhost' || host.endsWith('.localhost')) return false
  if (!host.includes('.') && !host.includes(':')) return false
  if (isPrivateOrReservedIp(host)) return false
  return true
}

/**
 * Strict hostname validation for apartments.com listing URLs.
 * Substring checks (`url.includes('apartments.com')`) are unsafe because
 * they match evil.com/apartments.com and apartments.com.evil.com.
 */
export function isApartmentsComUrl(value: unknown): boolean {
  if (!isSafePublicHttpUrl(value)) return false
  const url = parseUrl(value)
  return url !== null && APARTMENTS_COM_HOSTS.has(url.hostname.toLowerCase())
}

File: p11-platform/services/data-engine/utils/url_safety.py
"""
Shared SSRF-safe URL validation for scraping targets.

All user-supplied or database-stored URLs that the data engine will fetch
must pass these checks before any network request is made.
"""

import ipaddress
from typing import Optional
from urllib.parse import urlparse

ALLOWED_SCHEMES = ("http", "https")

APARTMENTS_COM_HOSTS = frozenset({
    "apartments.com",
    "www.apartments.com",
})


def is_safe_public_url(url: Optional[str]) -> bool:
    """
    Return True only for http(s) URLs pointing at public hostnames.

    Rejects:
    - non-http(s) schemes (file://, gopher://, ftp://, etc.)
    - localhost and *.localhost
    - IP literals that are private, loopback, link-local, or reserved
    - URLs with embedded credentials (user:pass@host)
    """
    if not url or not isinstance(url, str):
        return False
    try:
        parsed = urlparse(url.strip())
    except ValueError:
        return False
    if parsed.scheme not in ALLOWED_SCHEMES:
        return False
    if parsed.username or parsed.password:
        return False
    host = (parsed.hostname or "").lower()
    if not host or host == "localhost" or host.endswith(".localhost"):
        return False
    if "." not in host:
        # Bare single-label hostnames (internal service names) are not allowed.
        return False
    try:
        address = ipaddress.ip_address(host)
    except ValueError:
        return True  # Hostname, not an IP literal.
    return not (
        address.is_private
        or address.is_loopback
        or address.is_link_local
        or address.is_reserved
        or address.is_multicast
        or address.is_unspecified
    )


def is_apartments_com_url(url: Optional[str]) -> bool:
    """
    Strict hostname validation for apartments.com listing URLs.

    The hostname must be exactly apartments.com or www.apartments.com;
    substring checks like ``'apartments.com' in url`` are not safe because
    they match hosts such as evil.com/apartments.com or apartments.com.evil.com.
    """
    if not is_safe_public_url(url):
        return False
    host = (urlparse(url.strip()).hostname or "").lower()
    return host in APARTMENTS_COM_HOSTS
